### PR TITLE
weak_ptr - changes due to code review

### PR DESCRIPTION
**Pull request by majkel84**

### CODE REVIEW

File: homework/shared_ptr/weak_ptr.hpp [hunk 1]
```diff
@@ -29,6 +29,7 @@ class weak_ptr {
     T* ptr_{nullptr};
 
     void checkAndDeletePointers();
+    void checkWeakAndDeletePointers();
```
Comment (J-Cieplinski) [inline homework/shared_ptr/weak_ptr.hpp:32]: remove that

File: homework/shared_ptr/weak_ptr.hpp [hunk 2]
```diff
@@ -37,8 +38,21 @@ void weak_ptr<T>::checkAndDeletePointers() {
...
 
+template <typename T>
+void weak_ptr<T>::checkWeakAndDeletePointers() {
+    if (counter_) {
+        counter_->decreaseWeakRefs();
+        if (counter_->getRefs() == 0 && counter_->getWeakRefs() == 0) {
+            delete counter_;
+        }
+    }
+}
```
Comment (J-Cieplinski) [inline homework/shared_ptr/weak_ptr.hpp]: as well as this. It's basically duplicating `checkAndDeletePointers` and I want counter decrease to be explicitly called

File: homework/shared_ptr/weak_ptr.hpp [hunk 3]
```diff
@@ -72,6 +86,10 @@ weak_ptr<T>::~weak_ptr() {
 
 template <typename T>
 weak_ptr<T>& weak_ptr<T>::operator=(const weak_ptr& ptr) noexcept {
+    if (this == &ptr) {
+        return *this;
+    }
+    checkWeakAndDeletePointers();
```
Comment (J-Cieplinski) [inline homework/shared_ptr/weak_ptr.hpp]:
```suggestion
if(counter_) {
    counter_->decreaseWeakRefs();
    checkAndDeletePointers();
}
```

File: homework/shared_ptr/weak_ptr.hpp [hunk 4]
```diff
@@ -82,6 +100,8 @@ weak_ptr<T>& weak_ptr<T>::operator=(const weak_ptr& ptr) noexcept {
 
 template <typename T>
 weak_ptr<T>& weak_ptr<T>::operator=(const shared_ptr<T>& ptr) noexcept {
+    checkWeakAndDeletePointers();
```
Comment (J-Cieplinski) [inline homework/shared_ptr/weak_ptr.hpp]:
```suggestion
if(counter_) {
    counter_->decreaseWeakRefs();
    checkAndDeletePointers();
}
```

**Review by J-Cieplinski — CHANGES_REQUESTED** [review]

File: homework/shared_ptr/weak_ptr.hpp (hunk 4 above)
Comment (majkel84) [reply homework/shared_ptr/weak_ptr.hpp]: done

File: homework/shared_ptr/weak_ptr.hpp (hunk 3 above)
Comment (majkel84) [reply homework/shared_ptr/weak_ptr.hpp]: done

File: homework/shared_ptr/weak_ptr.hpp [hunk 5]
```diff
@@ -72,6 +73,13 @@ weak_ptr<T>::~weak_ptr() {
...
+    if (this == &ptr) {
+        return *this;
+    }
+    if(counter_) {
+        counter_->decreaseWeakRefs();
+        checkAndDeletePointers();
+    }
     ptr_ = ptr.ptr_;
     counter_ = ptr.counter_;
     if (counter_) {
```
Comment (J-Cieplinski) [inline homework/shared_ptr/weak_ptr.hpp]:
```suggestion
    if (this != &ptr) {
        if(counter_) {
            counter_->decreaseWeakRefs();
            checkAndDeletePointers();
        }
    ptr_ = ptr.ptr_;
    counter_ = ptr.counter_;
```

**Review by J-Cieplinski — APPROVED** [review]

**Pull request merged**